### PR TITLE
[feature] Support udf through RPC

### LINKED ISSUE #7520: [Feature] Support UDF through RPC
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Support UDF is realized through GRPC protocol. This brings several benefits:

The udf implementation language is not limited to c++, users can use any familiar language to implement udf
UDF is decoupled from Doris, udf will not be doris coredump together, udf computing resources are separated from doris, and doris services are not affected
But RPC's UDF has a fixed overhead, so its performance is much slower than C++ UDF, especially when the amount of data is large.

Create function like

CREATE FUNCTION rpc_add(INT, INT) RETURNS INT PROPERTIES (
  "SYMBOL"="add_int",
  "OBJECT_FILE"="127.0.0.1:9999",
  "MD5"=""
function service need to implements check_fn and fn_call methods

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by yangzhg**
## Proposed changes

Support UDF is implemented through GRPC protocol. This brings several benefits:

1.  The udf implementation language is not limited to c++, users can use any familiar language to implement udf
2. UDF is decoupled from Doris, udf will not be doris coredump together, udf computing resources are separated from doris, and doris services are not affected

But RPC's UDF has a fixed overhead, so its performance is much slower than C++ UDF, especially when the amount of data is large.

Create function like

```
CREATE FUNCTION rpc_add(INT, INT) RETURNS INT PROPERTIES (
  "SYMBOL"="add_int",
  "OBJECT_FILE"="127.0.0.1:9999",
  "TYPE"="RPC"
);
```

function service need to implements `check_fn` and `fn_call` methods
#7502 #7578 

the rpc udf call is  a little slower than  the native function call in vectorized mode:

image
image
But it will much slow in non-vectorized mode, and if using  rpc udf  in vectorized mode the query time is equal to native function call in non-vectorized mode
image

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #7520) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by stalary** [comment]
Let me see

File: be/src/exec/tablet_sink.cpp [hunk 1]
```diff
@@ -146,10 +147,8 @@ void NodeChannel::_cancel_with_msg(const std::string& msg) {
 Status NodeChannel::open_wait() {
     _open_closure->join();
     if (_open_closure->cntl.Failed()) {
-        if (!ExecEnv::GetInstance()->brpc_stub_cache()->available(_stub, _node_info.host,
-                                                                  _node_info.brpc_port)) {
-            ExecEnv::GetInstance()->brpc_stub_cache()->erase(_open_closure->cntl.remote_side());
-        }
+        ExecEnv::GetInstance()->brpc_internal_client_cache()->erase(
```
Comment (stalary) [inline be/src/exec/tablet_sink.cpp:150]: Why is it no longer necessary to judge unavailabe

File: be/src/exprs/rpc_fn_call.cpp [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    PArgs* args = request.mutable_args();
+    for (int i = 0; i < _children.size(); ++i) {
+        PParameter* param = args->add_values();
+        void* src_slot = context->get_value(_children[i], row);
+        PParameterType* ptype = param->mutable_type();
+        if (src_slot == nullptr) {
+            ptype->set_type(PParameterType::NULL_TYPE);
+            continue;
+        }
+        switch (_children[i]->type().type) {
```
Comment (stalary) [inline be/src/exprs/rpc_fn_call.cpp:84]: Whether need to support `TYPE_STRING`?

**Comment by stalary** [comment]
Whether RPC addresses support load balance?

**Comment by stalary** [comment]
_brpc_stub_cache split _internal_client_cache& _function_client_cache Whether compatibility problems exist

File: be/src/exprs/rpc_fn_call.cpp (hunk 2 above)
Comment (yangzhg) [reply be/src/exprs/rpc_fn_call.cpp:84]: Yes! I will  add later

File: be/src/exec/tablet_sink.cpp (hunk 1 above)
Comment (yangzhg) [reply be/src/exec/tablet_sink.cpp:150]: it is not cost so much when create a new stub

File: be/src/exprs/rpc_fn_call.cpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    _fn_context_index = context->register_func(state, return_type, arg_types, 0);
+
+    // _fn.scalar_fn.symbol
+    _rpc_function_symbol = _fn.scalar_fn.symbol;
+
+    _client = state->exec_env()->brpc_function_client_cache()->get_client(_fn.hdfs_location);
+
+    if (_client == nullptr) {
+        return Status::InternalError(
+                fmt::format("rpc env init error: {}/{}", _fn.hdfs_location, _rpc_function_symbol));
```
Comment (morningman) [inline be/src/exprs/rpc_fn_call.cpp:55]: `hdfs_location`?

File: be/src/exprs/rpc_fn_call.cpp [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    }
+    return Status::OK();
+}
+
+template <typename T>
+T RPCFnCall::interpret_eval(ExprContext* context, TupleRow* row) {
+    T res_val;
+    PFunctionCallResponse response;
+    Status st = _eval_children(context, row, &response);
+    if (!st.ok() || response.status().status_code() != 0 ||
```
Comment (morningman) [inline be/src/exprs/rpc_fn_call.cpp:226]: The situation of `response.status().status_code() != 0` has been checked in `_eval_children()`.
No need to check again here.

File: be/src/exprs/rpc_fn_call.cpp [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        (response.result().has_null() && response.result().null_map(0))) {
+        res_val.is_null = true;
```
Comment (morningman) [inline be/src/exprs/rpc_fn_call.cpp:228]: Is it ok to return null if RPC failed?

File: be/src/vec/functions/function_rpc.cpp [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+    }
+}
+
+Status RPCFnCall::execute(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                          size_t result, size_t input_rows_count, bool dry_run) {
+    PFunctionCallRequest request;
+    PFunctionCallResponse response;
+    request.set_function_name(_symbol);
+    int64_t name_hash = 0;
+    murmur_hash3_x64_64(_symbol.c_str(), _symbol.size(), 21217891, &name_hash);
```
Comment (morningman) [inline be/src/vec/functions/function_rpc.cpp:513]: Does `murmur_hash3_x64_64` work for ARM64?

File: gensrc/proto/types.proto [hunk 7]
```diff
@@ -63,3 +67,150 @@ message PUniqueId {
     required int64 lo = 2;
 };
 
+message PGenericType {
```
Comment (morningman) [inline gensrc/proto/types.proto:70]: This is a same datatype define in `data.proto`, how about unify them?
I will do this in PR #7939

File: gensrc/proto/types.proto [hunk 8]
```diff
@@ -63,3 +67,150 @@ message PUniqueId {
...
+    optional float float_value = 3;
+    optional int32 int32_value = 4;
+    optional int64 int64_value = 5;
+    optional uint32 uint32_value = 6;
+    optional uint64 uint64_value = 7;
+    optional bool bool_value = 8;
+    optional string string_value = 9;
+    optional bytes bytes_value = 10;
+    optional PDateTime datetime_value = 11;
+    optional bool is_null  = 12 [default = false];
```
Comment (morningman) [inline gensrc/proto/types.proto:156]: better to move `is_null` to the second place? (order 2)

File: gensrc/proto/types.proto [hunk 9]
```diff
@@ -63,3 +67,150 @@ message PUniqueId {
...
+    repeated float float_value = 3;
+    repeated int32 int32_value = 4;
+    repeated int64 int64_value = 5;
+    repeated uint32 uint32_value = 6;
+    repeated uint64 uint64_value = 7;
+    repeated bool bool_value = 8;
+    repeated string string_value = 9;
+    repeated bytes bytes_value = 10;
+    repeated PDateTime datetime_value = 11;
+    repeated bool null_map = 12;
```
Comment (morningman) [inline gensrc/proto/types.proto:171]: move `null_map` and `has_null` to the second and third place?

File: gensrc/proto/types.proto (hunk 7 above)
Comment (yangzhg) [reply gensrc/proto/types.proto:70]: That's a good idea, but I think the datatype define in `data.proto` can not describe complex types very well, `PGenericType` + `PList` + `PMap` + `PField` +  `PStruct` + `PDecimal` may be better. And BTW,  datatype should not  to complex, it should be used more easier between languages

**Review by morningman — APPROVED** [review]
LGTM

**Comment by morningman** [comment]
Document need to be added later

**Pull request merged**